### ISSUE TITLE
Code Lab Feedback

### ISSUE THREAD

**Issue by Artsen** [Artsen/codelab-css-intro#1]
Hey @chrisvfritz, can you take a look at this? It's [hosted here](http://artsen-lesson04.surge.sh/) and meets the following [project](http://lansingcodelab.com/lessons/css-intro/1) criteria:
- [x] Include at least one instance of the `margin` property.
- [x] Include at least one instance of the `padding` property.
- [x] Include at least one instance of the `border` property.
- [x] Include at least one instance of the `color` property.
- [x] Include at least one instance of the `background` property.
- [x] Include at least one instance of the `font-size` property.
- [x] Include at least one instance of the `line-height` property.
- [x] All properties must have a visible effect.
- [x] Return no errors (warnings are OK) when the HTML document is uploaded to the [W3C Markup Validation Service](http://validator.w3.org/#validate_by_upload)
- [x] Return no errors or warnings when contents of your CSS are pasted into [CSS Lint](http://csslint.net/)

<hr>

I used [this resource](http://www.base64css.com/) for the base64 data uri, I hope that is alright. Also, do you know of a way I could figure out the base64 myself (without using outside tools)?

**Comment by chrisvfritz** [comment]
Looking very nice! :shipit: And there's nothing wrong here, but I'll answer your question and use this as an opportunity to talk about a couple other tricks. :smile_cat:

### Base64 Data URIs

It's definitely alright that you base64 encoded the image! In simple apps like this, I'll usually just link to the image file itself instead - especially for larger images, where there's less of a benefit in minimizing the number of requests to the server.

In more complicated apps... well, things get more complicated. :wink: I typically use a build tool called Webpack to automatically base64 encode any images under a certain size that I link to in my HTML, CSS, or JavaScript. Webpack is notoriously unfriendly to newcomers though. 

### Unitless `line-height`s

While a `px` value for `line-height` _is_ technically valid, it's [generally recommended](https://css-tricks.com/almanac/properties/l/line-height/) to use the unitless equivalent instead. Without a unit, `line-height` works like line spacing in Word. So `1` is single space, `2` is double space, etc. I'm personally a huge fan of the `1.3` to `1.6` range. :smiley: 

Here's the full, technically explanation from [CSS Tricks](https://css-tricks.com/almanac/properties/l/line-height/), explaining exactly why this is recommended (hopefully it makes sense!):

> Unitless line heights are recommended due to the fact that child elements will inherit the raw number value, rather than the computed value. With this, child elements can compute their line heights based on their computed font size, rather than inheriting an arbitrary value from a parent that is more likely to need overriding.

### Using dashes instead of underscores for CSS classes

Using underscores for CSS classes is perfectly valid. _However_, if you use dashes instead, then there's [a special CSS selector](https://www.w3.org/Style/CSS/Test/CSS3/Selectors/current/html/full/flat/css3-modsel-8.html) you can use that only works with dashes. If you wanted to select anything that had the class `panel`, `panel-info`, or `panel-warning` for example, then you could use:

``` css
p[class|="panel"]
```

That symbol before the `=` is a pipe, available on the same key as the backslash on many keyboards.

**Comment by Artsen** [comment]
Very cool and wonderful info. I will make sure to save this for later.

It does seem that the dashes is better, and thanks for letting me know about the pipe trick! Do you know if that is mostly cross browser compatible? I know often times I run into clients who still use older forms of internet explorer who run into issues when using newer properties and values. Also I will indeed use unitless line-heights from now on as they seem like it would be better to organize rather than px or even worse em (which I don't even fully understand to be honest haha).

Thanks :D

**Comment by chrisvfritz** [comment]
Yep, the pipe trick has been [widely supported](http://caniuse.com/#feat=css-sel2) for a while now - just lesser known. 

And to be honest with you, I think _most_ people don't fully understand `em` and `rem`, so I don't think that's anything to feel bad about! If you're curious though, here's [the best article I've read](https://j.eremy.net/confused-about-rem-and-em/) about the subject - hopefully it may shed some light on those slippery units. :smiley:
